diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 0309873..4488d17 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -51,7 +51,6 @@ target_link_libraries(projectMSDL
         libprojectM::playlist
         Poco::Util
         SDL2::SDL2$<$<STREQUAL:${SDL2_LINKAGE},static>:-static>
-        SDL2::SDL2main
         )
 
 if(MSVC)
diff --git a/ImGui.cmake b/ImGui.cmake
index f578fc6..5e49779 100644
--- a/ImGui.cmake
+++ b/ImGui.cmake
@@ -57,6 +57,5 @@ target_link_libraries(ImGuiDemo
         PRIVATE
         ImGui
         SDL2::SDL2
-        SDL2::SDL2main
         OpenGL::GL
         )
