diff --git a/src/bumblebeed.c b/src/bumblebeed.c
index a911da9..35ba900 100644
--- a/src/bumblebeed.c
+++ b/src/bumblebeed.c
@@ -462,7 +462,7 @@ int main(int argc, char* argv[]) {
   bbconfig_parse_opts(argc, argv, PARSE_STAGE_PRECONF);
 
   /* First look for an intel card */
-  struct pci_bus_id *pci_id_igd = pci_find_gfx_by_vendor(PCI_VENDOR_ID_INTEL, 0);
+  struct pci_bus_id *pci_id_igd = pci_find_gfx_by_vendor(PCI_VENDOR_ID_ATI, 0);
   if (!pci_id_igd) {
     /* This is no Optimus configuration. But maybe it's a
        dual-nvidia configuration. Let us test that.
diff --git a/src/pci.h b/src/pci.h
index 476c3a1..2e951d9 100644
--- a/src/pci.h
+++ b/src/pci.h
@@ -21,6 +21,7 @@
 #pragma once
 #include <sys/types.h> /* necessary for int32_t */
 
+#define PCI_VENDOR_ID_ATI     0x1002
 #define PCI_VENDOR_ID_NVIDIA  0x10de
 #define PCI_VENDOR_ID_INTEL   0x8086
 #define PCI_CLASS_DISPLAY_VGA 0x0300
